fdbclient/include/fdbclient/BlobCipher.h

```diff
@@ -181,6 +191,176 @@ struct hash<BlobCipherDetails> {
 };
 } // namespace std
 
+#pragma pack(push, 1) // exact fit - no padding
+struct BlobCipherEncryptHeaderFlagsV1 {
+	constexpr static FileIdentifier file_identifier = 8269919;
+
+	// Serializable fields
+
+	uint8_t encryptMode;
+	uint8_t authTokenMode;
+	uint8_t authTokenAlgo;
+
+	bool operator==(const BlobCipherEncryptHeaderFlagsV1& o) const {
+		return encryptMode == o.encryptMode && authTokenMode == o.authTokenMode && authTokenAlgo == o.authTokenAlgo;
+	}
+
+	static void deserialize(const StringRef& flagsRef, BlobCipherEncryptHeaderFlagsV1& ret) {
+		ObjectReader headerReader(flagsRef.begin(), Unversioned());
+		headerReader.deserialize<BlobCipherEncryptHeaderFlagsV1>(ret);
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, encryptMode, authTokenMode, authTokenAlgo);
+	}
+};
+
+// Encryption header is stored as plaintext on a persistent storage to assist reconstruction of cipher-key(s)
+// for reads. FIPS compliance recommendation is to leverage cryptographic digest mechanism to generate
+// 'authentication token' (crypto-secure) to protect against malicious tampering and/or bit rot/flip scenarios.
+//
+// Encryption header support two modes of generation 'authentication tokens':
+// 1) SingleAuthTokenMode: the scheme generates single crypto-secrure auth token to protect {cipherText +
+// header} payload. Scheme is geared towards optimizing cost due to crypto-secure auth-token generation,
+// however, on decryption client needs to be read 'header' + 'encrypted-buffer' to validate the 'auth-token'.
+// The scheme is ideal for usecases where payload represented by the encryptionHeader is not large and it is
+// desirable to minimize CPU/latency penalty due to crypto-secure ops, such as: CommitProxies encrypted inline
+// transactions, StorageServer encrypting pages etc.
+// SOMEDAY: Another potential scheme could be 'MultiAuthTokenMode': Scheme generates separate authTokens
+// for 'encrypted buffer' & 'encryption-header'. The scheme is ideal where payload represented by
+// encryptionHeader is large enough such that it is desirable to optimize cost of upfront reading full
+// 'encrypted buffer', compared to reading only encryptionHeader and ensuring its sanity; for instance:
+// backup-files.
+
+template <uint32_t S>
+struct AesCtrWithAuthV1 {
+	constexpr static FileIdentifier file_identifier = 3110923;
```

Comment:
I thought the plan is to binary-serialize the algo-specific headers. Do we have numbers of what's the object-serialized encryption header size for each algorithm?

Reply:
And also if we choose to object serialize the algo-specific headers, versioning for them would be really redundant. We can either evolve the schema via flatbuffer or via adding a new algo. We probably don't need 3 ways to evolve the schema.